### ISSUE TITLE
`adaptive_relaxed_count` cost is set by the largest rate in a `vmap`ed batch, not the typical one

### ISSUE THREAD

**Issue by esennesh** [esennesh/numpyro#1]
## Summary

`_accumulate`'s horizon is `Theta(lambda)` by design — the docstring says so, and
that is fine. But `_core` runs it under `jax.vmap`, and a vmapped `lax.while_loop`
runs until *every* lane's predicate is false. So the trip count for the whole
batch is set by its **largest** rate.

One runaway coordinate therefore costs as much as if every coordinate had run away.
On 4096 lanes:

| batch | time per call |
| --- | ---: |
| all 4096 lanes at `rate=1e-3` | 0.5 ms |
| all 4096 lanes at `rate=1e4` | 754.7 ms |
| **one lane at `1e4`, other 4095 at `1e-3`** | **765.1 ms** |

A single bad lane in 4096 is a **1500x** slowdown, indistinguishable from all of
them being bad.

## Reproduction

```python
import time, jax, jax.numpy as jnp
import numpyro.distributions as dist
from numpyro.contrib.diag_sgd import adaptive_relaxed_count

u = jax.random.uniform(jax.random.key(0), (4096,))
f = jax.jit(lambda r, u: adaptive_relaxed_count(dist.Poisson(r), u, 0.01))

def timeit(r, tag):
    z = f(r, u); jax.block_until_ready(z)              # compile
    t = time.perf_counter()
    for _ in range(3):
        z = f(r, u)
    jax.block_until_ready(z)
    print(f"{tag:<42}{(time.perf_counter() - t) / 3 * 1000:8.1f} ms")

timeit(jnp.full(4096, 1e-3), "all lanes at 1e-3")
timeit(jnp.full(4096, 1e4), "all lanes at 1e4")
timeit(jnp.full(4096, 1e-3).at[0].set(1e4), "one lane at 1e4, rest at 1e-3")
```

Rate scaling on its own, for reference (4x2000 lanes, all at the same rate):

```
rate=1e-3      0.9 ms
rate=1         1.5 ms
rate=10        4.1 ms
rate=100      17.3 ms
rate=1000    123.1 ms
rate=1e4    1103.0 ms
```

Linear in `lambda` past `lambda ~ 1`, as the documented `mode + O(sqrt(mode))`
horizon implies.

## Where

`numpyro/contrib/diag_sgd.py`:

- `_accumulate` (~line 443) — the `while_loop`, terminating on
  `p_last < tol * p_max`
- `_core` (~line 516) — `jax.vmap(one)(params_flat, u_flat)` at ~line 525
- `_ADAPT_TOL = 1e-7`, `_ADAPT_MAX_ITERS = 100_000` (~lines 440-441)

`_accumulate`'s docstring says termination is *"independent of `u`* (every lane
gets the same bounded horizon ~ mode + O(sqrt(mode)), so a vmapped batch has no
straggler)". That reasoning is correct with respect to `u`, and the property it
claims does hold across lanes that share a rate. It does not hold across lanes
with **different** rates, which is the usual case when the rates are per-coordinate
variational parameters rather than a fixed constant.

## Why it may not have shown up in testing

It needs *heterogeneous* rates with an outlier. Testing DSGD at sensible,
roughly uniform rates costs a handful of iterations per lane and is free. The
cost only appears when something upstream — an optimiser taking too large a
step, say — inflates a few coordinates while the rest stay small.

## How it surfaced

Fitting a two-layer Poisson sparse-coding model, ~54k latent coordinates per
image x 64 images x 16 samples, so ~55M lanes. A proximal-SVI step at too large
a penalty inflated a handful of coordinates. The rates stayed finite, so nothing
downstream flagged a divergence, and the cell ran for **9 hours without
finishing** — against ~10 minutes for its neighbours. The lanes were almost
entirely at `lambda ~ 1e-4`; the few runaway ones set the horizon for all of
them.

## Possible fixes

1. **Bucket or sort lanes by rate** before the `vmap`, so a straggler only drags
   its own bucket.
2. **Per-lane horizon under a masked `scan`** rather than a shared `while_loop`,
   so a finished lane stops contributing work rather than only stopping
   contributing *mass*.
3. **Document it**, if the behaviour is considered acceptable — the current
   docstring actively suggests the opposite, which is what cost the nine hours.
   A caller who knows can bound its own rates.

(3) alone would have been enough here.

## Workaround in use downstream

Abandoning the solve when any rate passes a threshold far above anything the
model legitimately produces, treating that cell as diverged rather than clamping
any rate — clamping would silently change the answer while letting a diverged run
continue. This is a property of the caller's iterates, though, not a fix to the
cost characteristic.

## Environment

jax 0.10.2, numpyro from `feature/diagonalization_sgd`, CUDA and CPU alike (the
numbers above are CPU; the 9-hour cell was on GPU).
